stl/inc/iterator

```diff
@@ -513,6 +513,501 @@ private:
+                || constructible_from<iter_value_t<_Iter>, iter_reference_t<_Iter>>) {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Variantish_state::_Holds_iter,
+            "common_iterator can only be dereferenced if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (is_pointer_v<_Iter> || _Has_op_arrow<_Iter>) {
+            return (_Val._Iterator);
+        } else if constexpr (is_reference_v<iter_reference_t<_Iter>>) {
+            auto&& _Tmp = *_Val._Iterator;
+            return _STD addressof(_Tmp);
+        } else {
+            return _Proxy{*_Val._Iterator};
+        }
+    }
+
+    common_iterator& operator++() {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Variantish_state::_Holds_iter,
+            "common_iterator can only be incremented if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        ++_Val._Iterator;
+        return *this;
+    }
+
+    decltype(auto) operator++(int) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Variantish_state::_Holds_iter,
+            "common_iterator can only be incremented if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (forward_iterator<_Iter>) {
+            common_iterator _Tmp = *this;
+            ++_Val._Iterator;
+            return _Tmp;
+        } else {
+            return _Val._Iterator++;
+        }
+    }
+
+    // clang-format off
+    template <class _OIter, sentinel_for<_Iter> _OSe>
+        requires sentinel_for<_Se, _OIter>
+    _NODISCARD friend bool operator==(const common_iterator& _Left, const common_iterator<_OIter, _OSe>& _Right) {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(
+            _Left._Val._Contains != _Variantish_state::_Nothing && _Right._Val._Contains != _Variantish_state::_Nothing,
+            "common_iterators can only be compared if both hold a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+
+        if (_Left._Val._Contains == _Variantish_state::_Holds_iter) {
+            if (_Right._Val._Contains == _Variantish_state::_Holds_iter) {
+                if constexpr (equality_comparable_with<_Iter, _OIter>) {
+                    return _Left._Val._Iterator == _Right._Val._Iterator;
+                } else {
+                    return true;
+                }
+            } else {
+                return _Left._Val._Iterator == _Right._Val._Sentinel;
+            }
+        } else {
+            if (_Right._Val._Contains == _Variantish_state::_Holds_iter) {
+                return _Left._Val._Sentinel == _Right._Val._Iterator;
+            } else {
+                return true;
+            }
+        }
+    }
+
+    // clang-format off
+    template <sized_sentinel_for<_Iter> _OIter, sized_sentinel_for<_Iter> _OSe>
+        requires sized_sentinel_for<_Se, _OIter>

```

Comment:
I see that this is in the Standard - I am surprised that nothing seems to require that `_Se` is a sized sentinel for `_Iter`. No change requested.

Reply:
Symmetrically, it's also not required that `_OSe` is a sized sentinel for `_OIter`. We *never*/can't compare sentinels - if these are `common_iterator`s over the same range, they must end in the same place; if not then why are you calling this function? - so we only require the capability to difference the two iterator types, or complementary iterator and sentinel types.

### Design Hat
Using a minimal constraint here is anomalous in the Ranges design: usually cross-type operations require that the left and right types each support the operation symmetrically. I honestly don't recall why we didn't do so in this case, I _do_ recall refining an earlier requirement that wanted to be minimal but wasn't. I'll chalk it up to not investing enough design time in this compatibility tool.